File: packages/node-sdk/src/resources/access-passes.ts
import { HttpClient } from '../http-client';
import {
  AccessPass,
  IssueAccessPassParams,
  UpdateAccessPassParams,
  ListAccessPassesParams,
} from '../types';

/**
 * AccessPasses resource for managing access passes
 */
export class AccessPasses {
  private http: HttpClient;

  constructor(httpClient: HttpClient) {
    this.http = httpClient;
  }

  /**
   * Issue a new access pass
   * @param params - Parameters for issuing the access pass
   * @returns The created access pass with installation URL
   */
  async issue(params: IssueAccessPassParams): Promise<AccessPass> {
    return this.http.post<AccessPass>('/v1/access-passes', params);
  }

  /**
   * List access passes with optional filtering
   * @param params - Optional parameters for filtering
   * @returns Array of access passes
   */
  async list(params?: ListAccessPassesParams): Promise<AccessPass[]> {
    const sigPayload: Record<string, any> = {};

    if (params?.templateId) {
      sigPayload.template_id = params.templateId;
    }
    if (params?.state) {
      sigPayload.state = params.state;
    }

    return this.http.get<AccessPass[]>('/v1/access-passes', sigPayload);
  }

  /**
   * Update an access pass
   * @param params - Parameters for updating the access pass
   * @returns The updated access pass
   */
  async update(params: UpdateAccessPassParams): Promise<AccessPass> {
    const { accessPassId, ...updateData } = params;
    return this.http.patch<AccessPass>(`/v1/access-passes/${accessPassId}`, updateData);
  }

  /**
   * Suspend an access pass
   * @param accessPassId - The ID of the access pass to suspend
   * @returns Success response
   */
  async suspend(accessPassId: string): Promise<{ success: boolean; message: string }> {
    return this.http.post(`/v1/access-passes/${accessPassId}/suspend`);
  }

  /**
   * Resume a suspended access pass
   * @param accessPassId - The ID of the access pass to resume
   * @returns Success response
   */
  async resume(accessPassId: string): Promise<{ success: boolean; message: string }> {
    return this.http.post(`/v1/access-passes/${accessPassId}/resume`);
  }

  /**
   * Unlink an access pass from the user's device
   * @param accessPassId - The ID of the access pass to unlink
   * @returns Success response
   */
  async unlink(accessPassId: string): Promise<{ success: boolean; message: string }> {
    return this.http.post(`/v1/access-passes/${accessPassId}/unlink`);
  }

  /**
   * Delete an access pass permanently
   * @param accessPassId - The ID of the access pass to delete
   * @returns Success response
   */
  async delete(accessPassId: string): Promise<{ success: boolean; message: string }> {
    return this.http.post(`/v1/access-passes/${accessPassId}/delete`);
  }
}

File: packages/python-sdk/wusul/resources/access_passes.py
"""AccessPasses resource for managing access passes."""
from typing import Any, Dict, List, Optional

from ..types import (
    AccessPass,
    IssueAccessPassParams,
    ListAccessPassesParams,
    UpdateAccessPassParams,
)


class AccessPasses:
    """AccessPasses resource for managing access passes."""

    def __init__(self, http_client: Any):
        """
        Initialize the AccessPasses resource.

        Args:
            http_client: The HTTP client instance
        """
        self.http = http_client

    def issue(self, params: IssueAccessPassParams) -> AccessPass:
        """
        Issue a new access pass.

        Args:
            params: Parameters for issuing the access pass

        Returns:
            The created access pass with installation URL
        """
        return self.http.post("/v1/access-passes", params)

    def list(self, params: Optional[ListAccessPassesParams] = None) -> List[AccessPass]:
        """
        List access passes with optional filtering.

        Args:
            params: Optional parameters for filtering

        Returns:
            Array of access passes
        """
        sig_payload: Dict[str, Any] = {}

        if params:
            if params.get("templateId"):
                sig_payload["template_id"] = params["templateId"]
            if params.get("state"):
                sig_payload["state"] = params["state"]

        return self.http.get("/v1/access-passes", sig_payload)

    def update(self, params: UpdateAccessPassParams) -> AccessPass:
        """
        Update an access pass.

        Args:
            params: Parameters for updating the access pass

        Returns:
            The updated access pass
        """
        access_pass_id = params.pop("accessPassId")
        return self.http.patch(f"/v1/access-passes/{access_pass_id}", params)

    def suspend(self, access_pass_id: str) -> Dict[str, Any]:
        """
        Suspend an access pass.

        Args:
            access_pass_id: The ID of the access pass to suspend

        Returns:
            Success response
        """
        return self.http.post(f"/v1/access-passes/{access_pass_id}/suspend")

    def resume(self, access_pass_id: str) -> Dict[str, Any]:
        """
        Resume a suspended access pass.

        Args:
            access_pass_id: The ID of the access pass to resume

        Returns:
            Success response
        """
        return self.http.post(f"/v1/access-passes/{access_pass_id}/resume")

    def unlink(self, access_pass_id: str) -> Dict[str, Any]:
        """
        Unlink an access pass from the user's device.

        Args:
            access_pass_id: The ID of the access pass to unlink

        Returns:
            Success response
        """
        return self.http.post(f"/v1/access-passes/{access_pass_id}/unlink")

    def delete(self, access_pass_id: str) -> Dict[str, Any]:
        """
        Delete an access pass permanently.

        Args:
            access_pass_id: The ID of the access pass to delete

        Returns:
            Success response
        """
        return self.http.post(f"/v1/access-passes/{access_pass_id}/delete")
